lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -340,28 +420,30 @@ def status(self, plugin: OpenSearchPlugin) -> PluginState:
 
         # The _user_request_to_enable comes first, as it ensures there is a relation/config
         # set, which will be used by _is_enabled to determine if we are enabled or not.
-        if not self._user_requested_to_enable(plugin) and not self._is_enabled(plugin):
-            return PluginState.DISABLED
-
-        if self._is_enabled(plugin):
-            return PluginState.ENABLED
-
+        try:
+            if not self._is_enabled(plugin):
+                return PluginState.DISABLED
+            elif self._user_requested_to_enable(plugin):
+                return PluginState.ENABLED
+        except (
+            OpenSearchKeystoreNotReadyYetError,
+            OpenSearchPluginMissingConfigError,
+        ):
+            # We are missing configs or keystore. Report the plugin is only installed
+            pass
         return PluginState.INSTALLED
 
     def _is_installed(self, plugin: OpenSearchPlugin) -> bool:
         """Returns true if plugin is installed."""
-        return plugin.name in self._installed_plugins()
+        return plugin.name in self._installed_plugins
 
     def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
         """Returns True if user requested plugin to be enabled."""
         plugin_data = ConfigExposedPlugins[plugin.name]
-        if not (
-            self._charm.config.get(plugin_data["config"], False)
-            or self._is_plugin_relation_set(plugin_data["relation"])
-        ):
-            # User asked to disable this plugin
-            return False
-        return True
+        return self._charm.config.get(plugin_data["config"], False) or (
+            plugin_data["relation_handler"]
+            and plugin_data["relation_handler"](self._charm).is_relation_set()
```

Comment:
doesn't this force all plugins to be enabled by relation?

Reply:
No, KNN for example does not depend on relation_handler:
```
    "opensearch-knn": {
        "class": OpenSearchKnn,
        "config": "plugin_opensearch_knn",
        "relation_handler": None,
    },
```
That 2nd check is false as `plugin_data["relation_handler"]==None`